source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -401,67 +270,38 @@ void *ns_dyn_mem_temporary_alloc(int16_t alloc_size)
 }
 
 #ifndef STANDARD_MALLOC
-static void ns_free_and_merge_next_sector(int *cur_sector, int size)
+static void ns_free_and_merge_with_adjacent_sectors(int *cur_sector, int size)
 {
-    int *end_sector;
-    int next_len;
-    end_sector = cur_sector;
-    end_sector += size + 1;
-    if (end_sector != heap_main_end) {
-        end_sector++;
-        next_len = *end_sector;
-
-        if (next_len > 0) {
-            //Convert End of len
-            end_sector--;
-        } else {
-            //Merge First to second
-            //Set First End
-            //Skip Data And sector start of len
-            //debug("Free Next\r\n");
-            next_len = -(next_len);
-            end_sector += (next_len + 1);
-            size += (next_len + 2);
-        }
+    // Theory of operation: Sector is always in form | Len | Data | Len |
+    // So we need to check length of previous (if current not heap start)
+    // and next (if current not heap end) sectors. Negative length means
+    // free memory so we can merge freed sector with those.
 
-    }
-    //Set new sector start & End length
-    *cur_sector = -(size);
-    *end_sector = -(size);
-}
-
-static void ns_merge_prev_sector(int *cur_sector)
-{
-    int *prev_sector = 0;
-    int prev_len, cur_len;
+    int *start = cur_sector;
+    int *end = cur_sector + size + 1;
+    int total_size = size;
 
     if (cur_sector != heap_main) {
-        prev_sector = cur_sector;
-        prev_sector--;
-        prev_len = *prev_sector;
-
-        if (prev_len < 0) {
-            //Merge First to second
-            //Set First End
-            //Skip Data And sector start of len
-            //debug("Merge Prev\r\n");
-            prev_len = -(prev_len);
-            prev_sector -= (prev_len + 1); // SET PREV Start
-
-            cur_len = *cur_sector;
-            cur_len = -cur_len; //Invert
-            cur_sector += (cur_len + 1); //New End pointer
-            prev_len += (cur_len + 2); // SET NEW LEN
-        } else {
-            prev_sector = 0;
+        cur_sector--;
+        if (*cur_sector < 0) {
+            total_size += (2 - *cur_sector);
+            start -= (2 - *cur_sector);
         }
-
+        cur_sector++;
     }
-    //Set new sector start & End length
-    if (prev_sector) {
-        *prev_sector = -(prev_len);
-        *cur_sector = -(prev_len);
+
+    if (end != heap_main_end) {
+        end++;
+        if (*end < 0) {
+            total_size += (2 - *end);
+            end += (1 - *end);
+        }else{
+            end--;
+        }
     }
+    *cur_sector = -total_size;
```

Comment:
Why write to *cur_sector? Either it's the same as start, in which case it's pointless, or it's in the middle of a larger reason - still pointless.